### PR TITLE
Feature `CadShape` node

### PR DESCRIPTION
**Pull request by ad-daniel**
**Description**
To address this issue: https://github.com/cyberbotics/webots/issues/4195, this PR creates a full node (previously a PROTO) for the importing of collada objects

- [x] Add fields:
  - [x] url as MFString
  - [x] ccw
  - [x] castShadows
  - [x] isPickable
- [x] add segmentation and range finder visibility
- [x] add downloader
- [x] prevent it from being used in bounding object
- [x] replace pbr tokenizer init by direct init (and fix crash)
- [x] remove colladashape proto related code
- [x] remove model import and VRML export
- [x] add test 
- [x] docs https://cyberbotics.com/doc/reference/colladashape?version=feature-collada-shape-node
- [x] changelog (with backwards compatibility notice ColladaShapes -> ColladaShape)

Following the discussion here https://github.com/cyberbotics/urdf2webots/issues/168, the node should be made more generic and support both wavefront and collada files

- [x] add obj support
- [x] add test
- [x] update docs

### CODE REVIEW

File: src/webots/nodes/WbAbstractAppearance.cpp
Comment (omichel) [inline src/webots/nodes/WbAbstractAppearance.cpp:57]: It is OK to call delete with a NULL argument (it will simply do nothing):
```suggestion
  delete mTextureTransform;
```

File: src/webots/nodes/WbPbrAppearance.cpp
Comment (omichel) [inline src/webots/nodes/WbPbrAppearance.cpp:83]:
```suggestion
  const float opacity = 1.0f;
```

File: src/webots/nodes/WbColladaShape.cpp
Comment (omichel) [inline src/webots/nodes/WbColladaShape.cpp:411]:
```suggestion
  const float opacity = 1.0f;
```

File: src/webots/nodes/WbPbrAppearance.cpp
Comment (omichel) [inline src/webots/nodes/WbPbrAppearance.cpp:175]: No need to test for NULL before deleting objects.

**Review by omichel — CHANGES_REQUESTED** [review]
I believe the approach is pretty good.
Maybe it would help a little bit to add a comment to the shallow constructors so that we understand that these constructors should be used only in the special case where we want to build a shallow version of the node.

File: src/webots/vrml/WbSFColor.hpp
Comment (omichel) [inline src/webots/vrml/WbSFColor.hpp:30]:
```suggestion
  explicit WbSFColor(double r, double b, double g) : mValue(WbRgb(r, g, b)) {}
```

File: src/webots/vrml/WbSFString.hpp
Comment (omichel) [inline src/webots/vrml/WbSFString.hpp:29]:
```suggestion
  explicit WbSFString(const QString &s) : mValue(s) {}
```

File: src/webots/nodes/WbColladaShape.cpp
Comment (omichel) [inline src/webots/nodes/WbColladaShape.cpp:199]:
```suggestion
  const unsigned int flags = aiProcess_ValidateDataStructure | aiProcess_Triangulate | aiProcess_GenSmoothNormals |
```

File: src/webots/nodes/WbColladaShape.cpp
Comment (omichel) [inline src/webots/nodes/WbColladaShape.cpp:273]:
```suggestion
        warn(tr("Mesh '%1' has more than 100'000 vertices, it is recommended to reduce the number of vertices.")
```

**Review by omichel — CHANGES_REQUESTED** [review]
Overall, it looks very good to me.
I will now test it.

**Review by omichel — COMMENTED** [review]
There is no example in the distribution using ColladaShape.
It would be nice to either modify an existing proto or world to use ColladaShape or add a new example.
Or maybe it's coming in #4194?

File: src/webots/nodes/WbAbstractAppearance.cpp
Comment (ad-daniel) [reply src/webots/nodes/WbAbstractAppearance.cpp:44]: These can't be const since they are passed to assimp which is supposed to replace the contents (only if it finds it)

**Comment by ad-daniel** [comment]
Basically there's 2 issues going on:
- On linux the cat model fails to open because assimp cannot decode the collada file (it contains empty tags such as `<subject/>` which it doesn't know how to handle). I believe msys installs assimp 5.2.3 whereas linux/mac install 5.0.1. I've upgraded to 5.2.3 on Linux as well and it seems the issue goes away (and save 70MB in the process, the compiled library is significantly smaller)
- The reason it looked like it wasn't working on windows is because the collada file contains `<transparency>0.0</transparency>` tag. In my implementation, to determine the transparency I check first for the presence of `AI_MATKEY_TRANSPARENCYFACTOR`, and if that fails I determine the transparency from the `AI_MATKEY_OPACITY` value, setting the transparency to `1 - opacity` (as far as I can tell, 90% of the times the latter is what is defined). The issue is that for reasons I don't know, that transparency in the collada file ends up being encoded as an opacity, so I ended up enforcing `1.0 - 0.0 = 1.0` as transparency, making the cat invisible. So I don't think there's an issue with the code, I don't see much else we can do other than report the behavior and hope that at some point it gets fixed, but until then some files might have the same behavior

**Comment by ad-daniel** [comment]
The node is now generic and supports both collada and wavefront. A test was added that fails if all of a sudden the mapping changes for either format. It also ensures that a node added using `CadShape` compared to one built manually using `Shape { appearance PBRAppearance {} geometry Mesh {}}` (having the same roughness, basecolor, ... parameters of course) match.
A direct comparison between a `CadShape { url "something.dae" }` and `CadShape { url "same_something.obj" }` was not added on purpose since although similar they hardly ever would match in practice (since for instance obj defines the basecolor from a mat file, whereas collada does it differently).

Concerning the transparency, I've looked in the assimp codebase and it's a bit messy there as well, the reason is that different software programs apply the specifications of the various file formats in a very inconsistent/incomplete way, so assimp attempts to do its best while remaining as close as possible to the official specifications. In particular, I'm not even certain they consider it to be a bug since according to the code, depending on the situation, the transparency is flipped on purpose.
Since querying just the AI_MATKEY_OPACITY has worked well for years (for instance when importing from `3d model import` menu, or more recently with ColladaShapes PROTO), I've reverted my implementation to that approach instead (compared to the approach I used in this PR where I queried several MATKEYs related to transparency, not just AI_MATKEY_OPACITY).

Concerning the issues raised here https://github.com/cyberbotics/webots/issues/4447, the scaling was solved in a PR targetting master and the remaining issue about the inconsistency in the orientation (i.e., exporting as X-forward and on Webots not being X-forward) I will address it on a separate PR since it might entail updating all worlds that rely on the Mesh node.

This PR is currently blocking: https://github.com/cyberbotics/webots/pull/4194 and https://github.com/cyberbotics/urdf2webots/pull/169.

**Comment by ad-daniel** [comment]
PS: there is no rush for the review.

File: docs/reference/cadshape.md
Comment (omichel) [inline docs/reference/cadshape.md:13]:
```suggestion
The [CadShape](#cadshape) node renders a Collada (.dae) or Wavefront (.obj) object imported from an external file.
```

File: docs/reference/cadshape.md
Comment (omichel) [inline docs/reference/cadshape.md:16]:
```suggestion
```

File: tests/rendering/worlds/cadshape_node.wbt
Comment (omichel) [inline tests/rendering/worlds/cadshape_node.wbt:344]:
```suggestion
}

```

**Review by omichel — CHANGES_REQUESTED** [review]
Since this PR is quite big, it would be nice to have a second review.

**Review by BenjaminDeleze — CHANGES_REQUESTED** [review]
Seems to works well! However, I have some remarks:

- `CadShape` is missing from `nodes-and-api-functions.md`
- While I had no problem to generate an animation/scene, I did not manage to make the streaming work. It seems to be a problem of textures path.
- Is it deliberate to remove the vrml export feature ? If it is the case, I think we can clean a lot of thing as (if I am not mistaken) the only exporter left is x3d.
- `Move viewpoint to object` does not work (if I remember correctly it could be because of the bounding sphere)

File: src/webots/nodes/WbPbrAppearance.cpp
Comment (ad-daniel) [reply src/webots/nodes/WbPbrAppearance.cpp:689]: I think it's better yes, because they have been instantiated "artificially" (i.e not from a tokenizer nor copied from an existing node) so most of the member variables of these nodes, for instance at the basenode level, are not defined. Hence although letting the "normal" writer do the export might technically be possible, it entails adding a ton of conditionals just about everywhere to prevent segmentation faults at every corner, which is not really a clean approach and definitely harder to maintain in the long term. Or at least, that was the rationale behind the choice

File: resources/web/wwi/nodes/WbCadShape.js
Comment (ad-daniel) [reply resources/web/wwi/nodes/WbCadShape.js:5]: The reason was that once webotsjs will support tessellation, it will be closer to a WbBaseNode than a WbGroup, but you're right, now it's definitely the opposite

File: resources/web/wwi/Parser.js
Comment (ad-daniel) [reply resources/web/wwi/Parser.js:566]: You're right, it's how ccw was encoded for indexedfacesets before the cleanup the recent cleanup of webotsjs and I forgot to update it accordingly on this branch, I'll do it now

**Comment by BenjaminDeleze** [comment]
When streaming a world with a `CadShape`, a folder with the textures of the `CadShape` is created. For example, when I stream the test world I got:
`$WEBOTS_HOME/textures/cube.jpg`

Maybe the `CadShape` images are put there instead of being in `localhost:1234/textures` and that is why the steaming viewer is not finding the textures. (Just a hypothesis)

File: src/webots/nodes/WbSphere.cpp
Comment (omichel) [inline src/webots/nodes/WbSphere.cpp:119]: What means  TOREM?

**Review by omichel — DISMISSED** [review]

**Comment by ad-daniel** [comment]
Apologies, I think there's more that can be cleaned up, I just forgot to put it back to draft

**Comment by ad-daniel** [comment]
I think the PR can already be reviewed, the world update test will take a while but locally it came up clean, so fingers crossed should be fine. Linux test suite failed again due to `check_urdf` command missing, which appears to be happening a lot the last few days, or maybe it's just my impression.

I've cleaned up the obvious referenced to vrml exporter. Given that there's multiple writers still (urdf, webots), I don't think many of the `isX3d` can be removed  yet, or at least, not safely since we don't do much, to my knowledge, in terms of testing the exporters. So I've cleaned up the obvious cases only, with several exporters at play and little in terms of tests, it'd be a bit of a gamble to go further at this moment.

**Comment by ad-daniel** [comment]
PS: I've also fixed an edge case when parsing .dae as when opening some of the collada files from the franka emika robot some of them triggered an assert.

File: src/webots/vrml/WbMFDouble.cpp
Comment (omichel) [inline src/webots/vrml/WbMFDouble.cpp:18]: This one is likely useless:
```suggestion
```

File: src/webots/vrml/WbMFInt.cpp
Comment (omichel) [inline src/webots/vrml/WbMFInt.cpp:18]:
```suggestion
```

File: src/webots/vrml/WbProtoList.cpp
Comment (omichel) [inline src/webots/vrml/WbProtoList.cpp:24]:
```suggestion
```

File: src/webots/vrml/WbProtoModel.cpp
Comment (omichel) [inline src/webots/vrml/WbProtoModel.cpp:30]:
```suggestion
```

File: src/webots/vrml/WbSingleValue.cpp
Comment (omichel) [inline src/webots/vrml/WbSingleValue.cpp:16]:
```suggestion
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Review by BenjaminDeleze — APPROVED** [review]
Thank you

**Review by omichel — APPROVED** [review]

**Pull request merged**
